crates/biome_cli/src/reporter/gitlab.rs

```diff
@@ -0,0 +1,243 @@
+
+        let total_diagnostics = payload.diagnostics.len();
+        for (index, biome_diagnostic) in payload.diagnostics.into_iter().enumerate() {
+            if biome_diagnostic.severity() >= payload.diagnostic_level {
+                if biome_diagnostic.tags().is_verbose() && !payload.verbose {
+                    continue;
+                }
+
+                let diagnostic = self.builder.gitlab_diagnostic(biome_diagnostic);
+                let mut content = serde_json::to_string(&diagnostic)?;
+                if index < total_diagnostics - 1 {
+                    content.push(',')
+                }
+
+                self.console.log(markup!({ content }));
+            }
+        }
+        self.console.log(markup!("]"));
+        Ok(())
+    }
+}
+
+pub(crate) struct GitLabDiagnosticBuilder {
+    /// The root of the Git repository.
+    /// Required for relativization of the reported absolute paths.
+    repository_root: Option<PathBuf>,
+    /// A set of fingerprints (unique identifiers) to prevent collisions.
+    fingerprints: HashSet<u64>,
+}
+
+impl GitLabDiagnosticBuilder {
+    pub fn new(repository_root: Option<PathBuf>) -> Self {
+        Self {
+            repository_root,
+            fingerprints: HashSet::new(),
+        }
+    }
+
+    /// Turns a biome diagnostic into one in the GitLab Code Quality report format.
+    pub fn gitlab_diagnostic(
+        &mut self,
+        value: biome_diagnostics::error::Error,
+    ) -> GitLabDiagnostic {
+        let check_name = self.check_name(&value).unwrap_or_default();
+        let path = self.path(&value).unwrap_or_default();
+        let line = self.line(&value).map_or(1, |line| max(line, 1));
+        let source_code = self.source_code(&value).unwrap_or_default();
+        let fingerprint = self
+            .ensure_fingerprint_uniqueness(self.fingerprint(&check_name, &path, &source_code), 0)
+            .to_string();
+
+        GitLabDiagnostic {
+            severity: GitLabSeverity::from(value.severity()),
+            description: PrintDescription(&value).to_string(),
+            check_name,
+            fingerprint,
+            location: Location {
+                path,
+                lines: Lines { begin: line },
+            },
+        }
+    }
+
+    /// Extracts the name of the category as the check name.
+    fn check_name(&self, value: &biome_diagnostics::error::Error) -> Option<String> {
+        let category = value.category()?;
+        Some(category.name().to_string())
+    }
+
+    /// Extracts the source code generating the diagnostic.
+    fn source_code(&self, value: &biome_diagnostics::error::Error) -> Option<String> {
+        let location = value.location();
+        let source_code = location.source_code?;
+        let span = location.span?;
+
+        Some(source_code.text[span].to_string())
+    }
+
+    /// Generates a fingerprint for a diagnostic.
+    fn fingerprint(&self, check_name: &String, path: &String, code: &String) -> u64 {
```

Comment:
It's usually better to use `&str` instead of `&String`. You can use `as_str()` to retrieve a `&str` from `String`